src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java

```diff
@@ -208,4 +213,61 @@ public String deleteControlledGcsBucket(
 
     return jobBuilder.submit();
   }
+
+  /**
+   * Users who are READERS on the workspace may only have READER  on a private resource.
+   * WRITER workspace users may have either READER or WRITER assigned to them.
+   */
+  private void validateUserWorkspaceRole() {
+
+  }
+
+  private void validateOnlySelfAssignmentForWorkspaceWriter(ControlledResource controlledResource, AuthenticatedUserRequest userRequest) {
+    if (!controlledResource.getAccessScope().equals(AccessScopeType.ACCESS_SCOPE_PRIVATE)) {
+      // No need to handle SHARED resources
+      return;
+    }
+    final boolean userIsWriterOnWorkspace = samService.isAuthorized(
+        userRequest.getRequiredToken(),
+        SamConstants.SAM_WORKSPACE_RESOURCE,
+        controlledResource.getWorkspaceId().toString(),
+        SamConstants.SAM_WORKSPACE_WRITE_ACTION);
+    if (userIsWriterOnWorkspace) {
+      // UserEmail field isn't always populated for requests going through Swagger, so get it the hard way
+      final String requestUserEmail = samService.getEmailFromToken(userRequest.getRequiredToken());
```

Comment:
if you want to speed this up, you could check if the userRequest has a non-null/empty email first and use that if it's available.

Reply:
zl: check self-assignment 1st. For now, that's the only thing allowed.

In future, letting owners reassign can be done, at cost of some complexity.

Reply:
TODO: do we remove the assignedUser field from the API? We can deprecate it for now.